In [1]:
#importowanie bibliotek
import pandas as pd
import numpy as np
import sympy as sp
from sympy.abc import x,y 

In [2]:
#wczytanie danych
df = pd.read_csv("dane.txt")

In [3]:
#usunięcie kropek w zmiennej Churn?
df['Churn?'] = df['Churn?'].replace({'True.': True, 'False.': False})

C:\Users\USER\AppData\Local\Temp\ipykernel_1452\2909254064.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Churn?'] = df['Churn?'].replace({'True.': True, 'False.': False})


In [4]:
#utworzenie nowej ramki danych df2 ze zmiennymi Churn? oraz CustServ Calls
df2 = df[['Churn?', 'CustServ Calls']].copy()

In [5]:
df2

,Churn?,CustServ Calls
0,False,1
1,False,1
2,False,0
3,False,2
4,False,3
...,...,...
3328,False,2
3329,False,3
3330,False,2
3331,False,2


In [6]:
#utworzenie nowej zmiennej CSC

def csc_kategoria(x):
    if x == 0 or x == 1:
        return 'male'
    elif x == 2 or x == 3 :
        return 'srednie'
    else:
        return 'wysokie'

df2.loc[:, 'CSC'] = df2['CustServ Calls'].apply(csc_kategoria)

In [7]:
df2

,Churn?,CustServ Calls,CSC
0,False,1,male
1,False,1,male
2,False,0,male
3,False,2,srednie
4,False,3,srednie
...,...,...,...
3328,False,2,srednie
3329,False,3,srednie
3330,False,2,srednie
3331,False,2,srednie


In [8]:
#utworzeie tabeli krzyżowej 
tabk = pd.crosstab(df2['Churn?'], df2['CSC'], margins = True)

In [9]:
tabk

CSC,male,srednie,wysokie,All
Churn?,,,,
False,1664,1057,129,2850
True,214,131,138,483
All,1878,1188,267,3333


In [10]:
#nowe zmienne wskaźnikowe
def csc_sr(x):
    if x == 2 or x == 3:
        return 1
    else:
        return 0

def csc_wys(x):
    if x >= 4:
        return 1
    else:
        return 0

df2.loc[:, 'CSC_sr'] = df2['CustServ Calls'].apply(csc_sr)
df2.loc[:, 'CSC_wys'] = df2['CustServ Calls'].apply(csc_wys)

In [11]:
df2.head(15)

,Churn?,CustServ Calls,CSC,CSC_sr,CSC_wys
0,False,1,male,0,0
1,False,1,male,0,0
2,False,0,male,0,0
3,False,2,srednie,1,0
4,False,3,srednie,1,0
5,False,0,male,0,0
6,False,3,srednie,1,0
7,False,0,male,0,0
8,False,1,male,0,0
9,False,0,male,0,0


In [12]:
male_true = tabk.loc[True, 'male']
male_false = tabk.loc[False, 'male']

In [13]:
OR_nis = male_true/male_false
print("Iloraz liczby klientów, którzy odbyli małą liczbę rozmów i przenieśli się do innego operatora przez liczbę klientów,\nktórzy odbyli małą liczbę rozmów i nie przenieśli się wynosi: ", OR_nis)

Iloraz liczby klientów, którzy odbyli małą liczbę rozmów i przenieśli się do innego operatora przez liczbę klientów,
którzy odbyli małą liczbę rozmów i nie przenieśli się wynosi:  0.12860576923076922


In [14]:
srednie_true = tabk.loc[True, 'srednie']
srednie_false = tabk.loc[False, 'srednie']

In [15]:
OR_sr = (srednie_true * male_false) / (male_true * srednie_false)
print("Iloraz iloczynu liczby klientów, którzy odbyli średnią liczbę rozmów i przenieśli się oraz liczby klientów, \nktórzy odbyli mało rozmów i nie przenieśli się przez iloczyn liczby klientów, którzy odbyli mało rozmów i przenieśli się oraz liczby klientów, \nktórzy odbyli średnią liczbę rozmów i nie przenieśli się wynosi: ", OR_sr)

Iloraz iloczynu liczby klientów, którzy odbyli średnią liczbę rozmów i przenieśli się oraz liczby klientów, 
którzy odbyli mało rozmów i nie przenieśli się przez iloczyn liczby klientów, którzy odbyli mało rozmów i przenieśli się oraz liczby klientów, 
którzy odbyli średnią liczbę rozmów i nie przenieśli się wynosi:  0.9636866815798548


In [16]:
wysokie_true = tabk.loc[True, 'wysokie']
wysokie_false = tabk.loc[False, 'wysokie']

In [17]:
OR_wys = (wysokie_true * male_false) / (male_true * wysokie_false)
print("Iloraz iloczynu liczby klientów, którzy odbyli wysoką liczbę rozmów i przenieśli się oraz liczby klientów, \nktórzy odbyli mało rozmów i nie przenieśli się przez iloczyn liczby klientów, którzy odbyli mało rozmów i przenieśli się oraz liczby klientów, \nktórzy odbyli wysoką liczbę rozmów i nie przenieśli się wynosi: ", OR_wys)

Iloraz iloczynu liczby klientów, którzy odbyli wysoką liczbę rozmów i przenieśli się oraz liczby klientów, 
którzy odbyli mało rozmów i nie przenieśli się przez iloczyn liczby klientów, którzy odbyli mało rozmów i przenieśli się oraz liczby klientów, 
którzy odbyli wysoką liczbę rozmów i nie przenieśli się wynosi:  8.318191697457074


In [18]:
#obliczanie b0, b1 i b2
b0 = np.log(OR_nis)
b1 = np.log(OR_sr)
b2 = np.log(OR_wys)

print("b0 = ln(OR_nis) wynosi: ", b0)
print("b1 = ln(OR_sr) wynosi: ", b1)
print("b2 = ln(OR_wys) wynosi: ", b2)

b0 = ln(OR_nis) wynosi:  -2.0510036063593025
b1 = ln(OR_sr) wynosi:  -0.03698905630978367
b2 = ln(OR_wys) wynosi:  2.118444887154835


In [19]:
#zapisanie funkcji logitowej 
x,y = sp.symbols('x y')
g = b0 + b1 * x + b2 * y
print("Funkcja logitowa g(X) = b0 + b1x + b2y, gdzie X=[x,y] ma postać: ", g)

Funkcja logitowa g(X) = b0 + b1x + b2y, gdzie X=[x,y] ma postać:  -0.0369890563097837*x + 2.11844488715484*y - 2.0510036063593


In [24]:
#prawdopodobieństwo przeniesienia 
m = sp.exp(b0 + b1 * x + b2 * y) / (1 + sp.exp(b0 + b1 * x + b2 * y))
print("Prawdopodobieństwo przeniesienia się wynosi: ")
m

Prawdopodobieństwo przeniesienia się wynosi: 


0.128605769230769*exp(-0.0369890563097837*x + 2.11844488715484*y)/(0.128605769230769*exp(-0.0369890563097837*x + 2.11844488715484*y) + 1)

In [21]:
#obliczenie prawdopodobienstwa
p_male = sp.N(sp.exp(b0)/(1 + sp.exp(b0)))
p_srednie = sp.N(sp.exp(b0 + b1)/(1 + sp.exp(b0 + b1)))
p_wysokie = sp.N(sp.exp(b0 + b2)/(1 + sp.exp(b0 + b2)))

print("Prawdopodobieństwo przeniesienia się, gdy klient odbył mało rozmów:", p_male)
print("Prawdopodobieństwo przeniesienia się, gdy klient odbył średnio rozmów:", p_srednie)
print("Prawdopodobieństwo przeniesienia się, gdy klient odbył dużo rozmów:", p_wysokie)

Prawdopodobieństwo przeniesienia się, gdy klient odbył mało rozmów: 0.113951011714590
Prawdopodobieństwo przeniesienia się, gdy klient odbył średnio rozmów: 0.110269360269360
Prawdopodobieństwo przeniesienia się, gdy klient odbył dużo rozmów: 0.516853932584270


In [22]:
#prawdopodobienstwo przeniesienia sie bez podzialu na liczbe rozmow
total = tabk.loc['All', 'All']
churn_true_count = tabk.loc[True, 'All']

In [23]:
p_churn = churn_true_count / total
print("Prawdopodobieństwo przeniesienia bez podziału na rozmowy:", p_churn)

Prawdopodobieństwo przeniesienia bez podziału na rozmowy: 0.14491449144914492


## Analiza tabeli krzyżowej
Tabela krzyżowa pokazuje rozkład liczby klientów, którzy odeszli (`True`) i pozostali (`False`) w zależności od kategorii rozmów (`CSC`).  

- Klienci z **małą liczbą rozmów** stanowią grupę referencyjną.  
- Klienci z **średnią liczbą rozmów** i **dużą liczbą rozmów** mają większą liczbę przypadków odejścia w stosunku do grupy referencyjnej.

## Wnioski

Na podstawie tabeli krzyżowej oraz dalszych obliczeń można zauważyć, że liczba rozmów
z biurem obsługi klienta ma związek z prawdopodobieństwem przeniesienia się klienta,
jednak zależność ta nie jest jednakowa dla wszystkich kategorii.

Iloraz szans dla klientów z małą liczbą rozmów wynosi 0.129, co oznacza, że w tej grupie
przeniesienia zdarzają się rzadko. Klienci, którzy rzadko kontaktują się z biurem obsługi,
zazwyczaj nie decydują się na zmianę operatora.

Dla klientów ze średnią liczbą rozmów iloraz szans wynosi 0.964, czyli jest bardzo zbliżony
do wartości 1. Oznacza to, że ryzyko przeniesienia w tej grupie jest praktycznie takie samo
jak w grupie klientów z małą liczbą rozmów. Różnice pomiędzy tymi dwiema grupami są bardzo
niewielkie i nie mają większego znaczenia praktycznego.

Z kolei w przypadku klientów z dużą liczbą rozmów iloraz szans wynosi 8.318. Jest to
bardzo wysoka wartość, która wskazuje, że klienci często kontaktujący się z biurem obsługi
klienta mają znacznie większą skłonność do przeniesienia się w porównaniu do klientów
z małą liczbą rozmów.

Współczynniki modelu logitowego potwierdzają te obserwacje. Wartość parametru b0 jest
ujemna, co oznacza niskie bazowe ryzyko przeniesienia dla klientów z małą liczbą rozmów.
Parametr b1, odpowiadający średniej liczbie rozmów, jest bardzo bliski zeru, co wskazuje,
że ta kategoria nie wpływa istotnie na zmianę ryzyka przeniesienia. Natomiast parametr b2
jest dodatni i wysoki, co potwierdza silny wpływ dużej liczby rozmów na wzrost ryzyka odejścia.

Obliczone prawdopodobieństwa przeniesienia pokazują, że około 11% klientów z małą
i średnią liczbą rozmów zdecydowało się na przeniesienie, natomiast w grupie klientów
z dużą liczbą rozmów odsetek ten przekracza 50%. Oznacza to, że istotny wzrost ryzyka
przeniesienia pojawia się dopiero przy dużej liczbie kontaktów z biurem obsługi klienta.

Dla porównania, prawdopodobieństwo przeniesienia obliczone bez podziału na liczbę rozmów
wynosi około 14.5%. Wartość ta jest wyższa niż w grupach z małą i średnią liczbą rozmów,
ale znacznie niższa niż w grupie klientów często kontaktujących się z obsługą.

Podsumowując, sama obecność kontaktów z biurem obsługi klienta nie musi oznaczać
zwiększonego ryzyka przeniesienia, jednak ich duża liczba jest wyraźnym sygnałem
niezadowolenia klienta i wiąże się z bardzo wysokim prawdopodobieństwem odejścia.
